### PR TITLE
feat: [M3-6785] - Firewall-NodeBalancer → Dev

### PR DESCRIPTION
**Pull request by carrillo-erik**
## Description 📝
This PR merges the `Firewall-NodeBalancer` feature branch into the `develop` branch.

## Major Changes 🔄
- Updated the following flows to allow users to attach/assign `NodeBalancers` or `Linodes` to `Firewalls`:
  - Create A Linode
  - Create A Firewall
  - Create A NodeBalancer
- Updated the tabs for Linode, Firewall, and NodeBalancer.
- Updated events landing page to include NodeBalancer details.

### Prerequisites
- Turn on the Firewall-Nodebalancer feature flag in dev tools
- To go through the workflow, have atleast:
   - 2 Linodes (without an assigned Firewall)
      - 'Linode_1'
      - 'Linode_2' 
   - 2 NodeBalancers (without an assigned Firewall)
      - 'NodeBalancer_1'
      - 'NodeBalancer_2'

### Verification steps 
- Make sure the commit history looks accurate
- Scan code and make sure its the right set of code, since the code has been reviewed before, we are more concerned that this PR encapsulates everything needed.
- Run through the feature as a user, you can use the following workflow as a reference:
   - Throughout this workflow, ensure that:
      - There are no grammatical errors
      - The Linode, NodeBalancer and Firewall links bring you to the right page
      - The action is being added to the events: http://localhost:3000/events
   - If you do not have 2 unassigned Linodes and 2 unassigned NodeBalancers create them here: [Linode](http://localhost:3000/linodes/create), [NodeBalancer](http://localhost:3000/nodebalancers/create)
   - First, create a new Firewall (lets call this 'Firewall_1') without assigning any services: http://localhost:3000/firewalls/create
   - Navigate to Firewall_1's Linodes details page - `/firewalls/[firewall_1_id]/linodes`
   - While doing the below steps, make sure the number to the right of the tab headers (ex. Linode<ins>(#)</ins> and NodeBalancer<ins>(#)</ins>) get updated accordingly
   - Assign a Linode (lets call this 'Linode_1') to Firewall_1 - `/firewalls/[firewall_1_id]/linodes/add`
   - Ensure the Linode's table gets populated with the new Linode_1 - `/firewalls/[firewall_1_id]/linodes`
   - Assign another Linode (lets call this 'Linode_2') to Firewall_1 - `/firewalls/[firewall_1_id]/linodes/add`
   - Test the table's search bar functionality at `/firewalls/[firewall_1_id]/linodes` with the two Linodes
   - Unassign Linode_2 using the `remove` button in the table, and ensure the Linode gets properly unassigned.
   - Navigate to Firewall_1's NodeBalancer's details page - `/firewalls/[firewall_1_id]/linodes`
   - Assign a NodeBalancer_1 to Firewall_1 - `/firewalls/[firewall_1_id]/nodebalancers/add`
   - Ensure the NodeBalancer's table get populated with the new NodeBalancer_1 - `/firewalls/[firewall_1_id]/nodebalancers`
   - Ensure NodeBalancer_1's Details page gets updated with Firewall_1 - `/nodebalancers/[nodebalancer_1_id]/networking` and `/nodebalancers/[nodebalancer_1_id]/settings`
   - Navigate back to `/firewalls/[firewall_1_id]/nodebalancers/add` to assign another NodeBalancer, NodeBalancer_2, to Firewall_1
   - Test the NodeBalancer's table's search bar functionality at `/firewalls/[firewall_1_id]/nodebalancers` with the two assigned NodeBalancers.
   -  Unassign NodeBalancer_2 using the `remove` button in the table, and ensure the NodeBalancer gets properly unassigned.
   - Create a new Firewall (lets call this 'Firewall_2') - http://localhost:3000/firewalls/create
   - Ensure you cannot assign the already assigned Linode_1 to Firewall_2 - `/firewalls/[firewall_2_id]/linodes/add` 
      - There is an open [PR](https://github.com/linode/manager/pull/9935) that filters out already assigned services in the dropdowns; however, if this is not mergd by the time you do the review, atleast make sure there is an error that occurs if you do try to add an already assigned service
   - But also ensure that you _can_ assign Linode_2 to Firewall_2, and make sure to assign it
   - Ensure you cannot assign the already assigned NodeBalancer_1 to Firewall_2 - `/firewalls/[firewall_2_id]/nodebalancers/add` 
   - But also ensure that you _can_ assign NodeBalancer_2 to Firewall_2, and make sure to assign it
   - Navigate back to the firewalls landing page - http://localhost:3000/firewalls
   - Ensure that there are two Firewalls in the table - 'Firewall_1' and 'Firewall_2'
   - Ensure that there is a 'Services' column
   - Ensure that each Firewall's Services' column has the corresponding Linode and NodeBalancer that was assigned to. that firewall: Firewall_1 --> Linode_1 and NodeBalancer_1, Firewall_2 --> Linode_2 and NodeBalancer_2
   - Ensure that swapping on and off the feature flag disables and enables all of the feature related ui found [here](https://github.com/linode/manager/pull/9931)

### CODE REVIEW

File: packages/manager/src/features/Firewalls/FirewallLanding/FirewallRow.styles.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  label: 'StyledDevicesLink',
+})(() => ({
+  display: 'inline-block',
+}));
+
+export const StyledTruncateLinks = styled('div', {
+  label: 'StyledTruncateLinks',
+})(() => ({
+  WebkitBoxOrient: 'vertical',
+  WebkitLineClamp: 3,
```
Comment (abailly-akamai) [inline packages/manager/src/features/Firewalls/FirewallLanding/FirewallRow.styles.ts:14]: Need an alternative here
https://developer.mozilla.org/en-US/docs/Web/CSS/box-orient

1. box-orient is deprecated
2. only supports webkit based browsers

File: packages/manager/src/features/NodeBalancers/NodeBalancerCreate.tsx [hunk 2]
```diff
@@ -59,6 +62,7 @@ import type { APIError } from '@linode/api-v4/lib/types';
 
 interface NodeBalancerFieldsState {
   configs: (NodeBalancerConfigFieldsWithStatus & { errors?: any })[];
+  firewall_id?: number;
```
Comment (abailly-akamai) [inline packages/manager/src/features/NodeBalancers/NodeBalancerCreate.tsx:65]: another way of writing `firewallId`??

File: packages/manager/src/features/NodeBalancers/NodeBalancerDetail/NodeBalancerFirewalls.tsx [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+import { TableRowLoading } from 'src/components/TableRowLoading/TableRowLoading';
+import { Typography } from 'src/components/Typography';
+import { RemoveDeviceDialog } from 'src/features/Firewalls/FirewallDetail/Devices/RemoveDeviceDialog';
+import { useNodeBalancersFirewallsQuery } from 'src/queries/nodebalancers';
+
+import { NodeBalancerFirewallsRow } from './NodeBalancerFirewallsRow';
+
+interface Props {
+  displayFirewallInfoText: boolean;
+  nodeBalancerID: number;
```
Comment (abailly-akamai) [inline packages/manager/src/features/NodeBalancers/NodeBalancerDetail/NodeBalancerFirewalls.tsx:24]: same comment on the ID

**Comment by tyler-akamai** [comment]
> I went through most of the verification steps so far and the functionality is mostly good. One issue I encountered was trying to filter on the Linodes and NodeBalancer tabs when I have multiple of either. For example, I entered a string to search for among the assigned Linodes, and couldn't get the full list back when I cleared the search bar:
> 
>  Screen.Recording.2023-11-30.at.4.48.30.PM.mov

Good catch, should be fixed now

**Review by abailly-akamai — COMMENTED** [review]
Nice work! this is a big one.

Did a first pass to review the code only (will do a second one to test the UI)

Left a lot of comments, some if not most should be addressed when possible

File: packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx [hunk 4]
```diff
@@ -13,15 +13,19 @@ import { Autocomplete } from '../Autocomplete/Autocomplete';
 import { LinkButton } from '../LinkButton';
 
 interface Props {
+  entityType?: FirewallDeviceEntityType;
```
[the remark this answers is not in the document]
Comment (tyler-akamai) [reply packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx:16]: The Create Firewall Drawer varies based on navigating from three different pages
1. Create Linode Flow - entityType='linode'
2. Create NodeBalancer Flow - entityType='nodebalancer'
3. Firewall Landing page - entityType=undefined

each of these corresponds to different text in the Create Firewall Drawer, so the context of where is came from needs to be passed in. And since there are three options, two of which of of type FirewallDeviceEntityType and the third is just the normal landing page, I thought this would be a good time to have an optional type.

File: packages/manager/src/features/Firewalls/FirewallLanding/FirewallDialog.tsx [hunk 5]
```diff
@@ -81,7 +81,7 @@ export const FirewallDialog = React.memo((props: Props) => {
       open={open}
       title={`${capitalize(mode)} Firewall ${label}?`}
     >
-      Are you sure you want to {mode} this Firewall?
+      Are you sure you want to {mode} this firewall?
```
[the remark this answers is not in the document]
Comment (tyler-akamai) [reply packages/manager/src/features/Firewalls/FirewallLanding/FirewallDialog.tsx:84]: idk if I should be using `useHistory` here instead of passing in props, thoughts?

[the remark this answers is not in the document]
Comment (abailly-akamai) [reply packages/manager/src/features/Firewalls/FirewallLanding/FirewallDialog.tsx:84]: Up to what you find the cleanest. If it requires you to prop drill more than a couple levels down then you may want to use the hook. It's pretty suggestive

**Comment by coliu-akamai** [comment]
really great job Tyler & Erik!
Went through the UI flow, things look good + matched mockups! Had a few comments below:

- Accessiblity:
  - When using voiceover, the description of the NodeBalancer table in a Firewall's detail page for the Nodebalancer tab is 'List of Linodes attached to this firewall' - probably due to line 61 of `FirewallDeviceTable.tsx`
  - If needed, maybe we could also add some aria-labels to the Firewall Landing table

File: packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (tyler-akamai) [reply packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx:16]: idk if I should be using useHistory here instead of passing in props, thoughts?

[the remark this answers is not in the document]
Comment (tyler-akamai) [reply packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx:16]: I like that solution! Thank you

File: packages/manager/package.json [hunk 6]
```diff
@@ -126,6 +126,7 @@
     "@testing-library/user-event": "^12.1.1",
     "@types/braintree-web": "^3.75.23",
     "@types/chart.js": "^2.9.21",
+    "@types/clamp-js": "^0.7.6",
```
Comment (tyler-akamai) [inline packages/manager/package.json:129]: This was the only way we found to clamp firewall links in the column. Previously we were using -web-line-clamp but its being deprecated. I also tried all of the CSS solutions found [here](https://css-tricks.com/line-clampin/) but none of them looked good.

File: packages/manager/src/features/Firewalls/FirewallLanding/FirewallRow.styles.ts (hunk 1 above)
Comment (tyler-akamai) [reply packages/manager/src/features/Firewalls/FirewallLanding/FirewallRow.styles.ts]: Great catch, I ended up using clamp-js.

This was the best solution I found. I also tried all of the CSS solutions found [here](https://css-tricks.com/line-clampin/) but none of them looked good.

Comment (abailly-akamai) [reply packages/manager/src/features/Firewalls/FirewallLanding/FirewallRow.styles.ts]: Let's pair on that - am hoping we can find a css/markup only solution before committing to a new library at it

Comment (abailly-akamai) [reply packages/manager/src/features/Firewalls/FirewallLanding/FirewallRow.styles.ts]: Also, you were trying to reference clamp-js but did not install it (and we don't want to). In the worst case (using a lib) we should use { clamp } from ramda

File: packages/manager/src/features/Firewalls/FirewallDetail/Devices/AddLinodeDrawer.tsx
[the remark this answers is not in the document]
Comment (tyler-akamai) [reply packages/manager/src/features/Firewalls/FirewallDetail/Devices/AddLinodeDrawer.tsx:1]: Just added some unit tests to cover basic functionality/rendering.

We will also be adding e2e tests to cover some additional functionality.

File: packages/manager/src/features/Firewalls/FirewallDetail/Devices/AddNodebalancerDrawer.tsx [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+    if (!firstError) {
+      onClose();
+    }
+  };
+
+  const errorNotice = () => {
+    let errorMsg = localError || '';
+    // match something like: NodeBalancer <nodebalancer_label> (ID <nodebalancer_id>)
+
+    const nodebalancer = /NodeBalancer (.+?) \(ID ([^\)]+)\)/i.exec(errorMsg);
```
[the remark this answers is not in the document]
Comment (tyler-akamai) [reply packages/manager/src/features/Firewalls/FirewallDetail/Devices/AddNodebalancerDrawer.tsx:107]: The errorMsg is from either an api response or a hardcoded value in the file, ex. 'Could not load NodeBalancer Data'.

I did sanitize it using the sanitize-html library.

let me know if this fixes the issue you addresses.

File: packages/manager/src/features/Events/eventMessageGenerator.ts [hunk 8]
```diff
@@ -29,6 +29,11 @@ export const safeSecondaryEntityLabel = (
   return label ? `${text} ${label}` : fallback;
 };
 
+const secondaryEntityTypeObj = {
+  linode: 'Linode',
+  nodebalancer: 'NodeBalancer',
+};
```
[the remark this answers is not in the document]
Comment (tyler-akamai) [reply packages/manager/src/features/Events/eventMessageGenerator.ts:40]: added a `SecondaryEntityType` interface

File: packages/manager/.changeset/pr-9926-upcoming-features-1700672014523.md [hunk 9]
```diff
@@ -0,0 +1,5 @@
+---
+"@linode/manager": Upcoming Features
+---
+
+Firewall NodeBalancer links not working ([#9926](https://github.com/linode/manager/pull/9926))
```
Comment (dwiley-akamai) [inline packages/manager/.changeset/pr-9926-upcoming-features-1700672014523.md:5]:
```suggestion
Added Firewall documentation links in the Create flows for Linodes, NodeBalancers, and Firewalls and in the Add NodeBalancer to Firewall drawer ([#9926](https://github.com/linode/manager/pull/9926))
```

**Review by dwiley-akamai — COMMENTED** [review]
Functionality looks good 👍 I noticed a couple of minor things:

When creating a NodeBalancer and assigning a firewall in that flow, should "Firewall Assigned" show up in the summary section?

Screenshot 2023-12-05 at 12 33 13 PM

I think we need a period after "Learn more" here:

<details>
<summary>Create Firewall drawer</summary>

Screenshot 2023-12-05 at 12 37 58 PM
</details>

**Comment by tyler-akamai** [comment]
> Functionality looks good 👍 I noticed a couple of minor things:
> 
> When creating a NodeBalancer and assigning a firewall in that flow, should "Firewall Assigned" show up in the summary section?
> 
> Screenshot 2023-12-05 at 12 33 13 PM
> 
> I think we need a period after "Learn more" here:
> 
> Create Firewall drawer

Really good catch, I broke it during a recent push when I was changing `firewallID` to `FirewallId` :(

couldnt figure out what the issue was until Banks jumped on a call and helped, Typescript didnt catch it for some reason 🤔

File: packages/manager/src/features/NodeBalancers/NodeBalancerCreate.tsx [hunk 10]
```diff
@@ -286,6 +294,16 @@ const NodeBalancerCreate = () => {
...
     createNodeBalancer(nodeBalancerRequestData)
       .then((nodeBalancer) => {
+        addDevice({ id: nodeBalancer.id, type: 'nodebalancer' }).then(
+          (service) => {
+            queryClient.invalidateQueries([
+              'nodebalancers',
+              'nodebalancer',
+              service.id,
+              'firewalls',
+            ]);
```
Comment (bnussman-akamai) [inline packages/manager/src/features/NodeBalancers/NodeBalancerCreate.tsx:304]: Is this invalidation needed? If I'm understanding this correctly, the NodeBalancer was just created. There shouldn't be any data in `['nodebalancers', 'nodebalancer', service.id, 'firewalls']` that needs invalidation

File: packages/manager/package.json (hunk 6 above)
Comment (bnussman-akamai) [reply packages/manager/package.json:133]: This is fine for now, but I'd really like to see if we can figure out a solution that does not involve an additional package in the future if possible

File: packages/manager/src/features/Linodes/LinodesDetail/LinodeNetworking/LinodeFirewalls/LinodeFirewallsRow.tsx [hunk 11]
```diff
@@ -43,9 +43,9 @@ export const LinodeFirewallsRow = (props: LinodeFirewallsRowProps) => {
       key={`firewall-${firewallID}`}
     >
       <TableCell data-qa-firewall-label>
-        <StyledLink tabIndex={0} to={`/firewalls/${firewallID}`}>
+        <Link tabIndex={0} to={`/firewalls/${firewallID}`}>
```
Comment (coliu-akamai) [inline packages/manager/src/features/Linodes/LinodesDetail/LinodeNetworking/LinodeFirewalls/LinodeFirewallsRow.tsx:46]: There's a lot of commonality between LinodeFirewallsRow.tsx and NodeBalancerFirewallsRow.tsx - could be worth trying to abstract the two, maybe in some other ticket down the line if it's not quick

File: packages/manager/src/features/NodeBalancers/NodeBalancerDetail/NodeBalancerFirewalls.tsx [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+      <NodeBalancerFirewallsRow
+        firewall={attachedFirewall}
+        key={`firewall-${attachedFirewall.id}`}
+        nodeBalancerID={nodeBalancerId}
+        onClickUnassign={handleClickUnassign}
+      />
+    ));
+  };
+
+  // TODO: NBFW - Add Documentation Link
```
Comment (coliu-akamai) [inline packages/manager/src/features/NodeBalancers/NodeBalancerDetail/NodeBalancerFirewalls.tsx:79]: has the documentation link been provided yet?

File: packages/manager/src/features/Firewalls/FirewallDetail/Devices/AddLinodeDrawer.test.tsx [hunk 13]
```diff
@@ -0,0 +1,34 @@
+import * as React from 'react';
```
[the remark this answers is not in the document]
Comment (tyler-akamai) [reply packages/manager/src/features/Firewalls/FirewallDetail/Devices/AddLinodeDrawer.test.tsx:1]: Yeah so I tried writing the tests, but in order for the 'Add' button to be enabled, you have to mock a Firewall Device and then select it in the dropdown, there was also discussion about it earlier about how it might be better and simpler as an e2e test: https://github.com/linode/manager/pull/9886#discussion_r1401090255

File: packages/manager/src/features/Firewalls/FirewallDetail/Devices/FirewallDeviceLanding.test.tsx [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+        );
+        const addButton = getByTestId('add-device-button');
+        const table = getByRole('table');
+
+        expect(addButton).toBeInTheDocument();
+        expect(table).toBeInTheDocument();
+      });
+
+      if (prop.disabled) {
+        it(`should contain a disabled Add ${serviceName} button`, () => {
```
Comment (mjac0bs) [inline packages/manager/src/features/Firewalls/FirewallDetail/Devices/FirewallDeviceLanding.test.tsx:57]:
```suggestion
        it('should contain a disabled Add ${serviceName} button', () => {
```

File: packages/manager/src/features/Firewalls/FirewallDetail/Devices/FirewallDeviceLanding.test.tsx [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+          const { getByRole } = renderWithTheme(
+            <FirewallDeviceLanding {...prop} />
+          );
+          const permissionNotice = getByRole('alert');
+          expect(permissionNotice).toBeInTheDocument();
+        });
+      }
+
+      if (!prop.disabled) {
+        it(`should contain an enabled Add ${serviceName} button`, () => {
```
Comment (mjac0bs) [inline packages/manager/src/features/Firewalls/FirewallDetail/Devices/FirewallDeviceLanding.test.tsx:75]:
```suggestion
        it('should contain an enabled Add ${serviceName} button', () => {
```

File: packages/manager/src/features/Firewalls/FirewallDetail/Devices/FirewallDeviceLanding.test.tsx [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+          const { getByTestId } = renderWithTheme(
+            <FirewallDeviceLanding {...prop} />
+          );
+          const addButton = getByTestId('add-device-button');
+
+          expect(addButton).toBeEnabled();
+        });
+        it(`should navigate to Add ${serviceName} To Firewall drawer when enabled`, () => {
```
Comment (mjac0bs) [inline packages/manager/src/features/Firewalls/FirewallDetail/Devices/FirewallDeviceLanding.test.tsx:83]:
```suggestion
        it('should navigate to Add ${serviceName} To Firewall drawer when enabled', () => {
```

**Review by mjac0bs — COMMENTED** [review]
This is a behemoth of a PR; thanks for the live walkthrough and keeping track of all the feedback from the team.

The remaining things that I found in code were minor and could be follow ups if needed:

- Let's please fix the typo `occured` and replace with `occurred` (two `r`s) in the error state of NodeBalancerSelect and LinodeSelect.
- We should clean up the UI for the error handling of the firewalls table on the NB details and Linode details pages when we get a chance.
- The NodeBalancer summary panel doesn't have the right firewall link color in dark mode.
- Basically everything else was just optional, nice-to-have code cleanup and you can accept the suggestions if you'd like.

File: packages/api-v4/src/nodebalancers/types.ts
Comment (bnussman-akamai) [inline packages/api-v4/src/nodebalancers/types.ts:1]: I'm not sure why there is not a Typescript error in `NodeBalancerCreate.tsx`, but I think `CreateNodeBalancerPayload` should have `firewall_id?: number` added to it.

File: packages/manager/.changeset/pr-9886-upcoming-features-1699461580236.md [hunk 17]
```diff
@@ -0,0 +1,5 @@
+---
+"@linode/manager": Upcoming Features
+---
+
+Dropdowns in Create Firewall Drawer using Autocomplete instead of updated LinodeSelect and NodeBalancerSelect ([#9886](https://github.com/linode/manager/pull/9886))
```
[the remark this answers is not in the document]
Comment (tyler-akamai) [reply packages/manager/.changeset/pr-9886-upcoming-features-1699461580236.md:5]: Yeah it was changed to the component selects, but I was trying to follow your changset description convention: https://github.com/linode/manager/pull/9812#discussion_r1364032040

[the remark this answers is not in the document]
Comment (tyler-akamai) [reply packages/manager/.changeset/pr-9886-upcoming-features-1699461580236.md:5]: So I described the bug by saying 'Dropdowns in Create Firewall Drawer using Autocomplete instead of ...'

but there seems to still be a misunderstanding, so we can meet next week about how to name them for parent/child

File: packages/manager/src/features/NodeBalancers/NodeBalancerDetail/NodeBalancerSummary/SummaryPanel.tsx [hunk 18]
```diff
@@ -90,6 +95,16 @@ export const SummaryPanel = () => {
...
         </StyledSummarySection>
       </StyledSummarySectionWrapper>
+      {displayFirewallLink && flags.firewallNodebalancer ? (
+        <StyledSummarySection>
+          <StyledTitle data-qa-title variant="h3">
+            Firewall
+          </StyledTitle>
+          <Link className="secondaryLink" to={`/firewalls/${linkID}`}>
+            {linkText}
+          </Link>
```
[the remark this answers is not in the document]
Comment (carrillo-erik) [reply packages/manager/src/features/NodeBalancers/NodeBalancerDetail/NodeBalancerSummary/SummaryPanel.tsx]: @mjac0bs Good eye! Both had the appropriate `secondaryLink` class name. The difference between the two was that the port link is wrapped in a `<Typography />` component, whereas the Firewall link was not. This has been fixed now.

abcd

**Review by mjac0bs — APPROVED** [review]
Thanks for addressing all that feedback. Confirmed table error state is fixed, typos corrected, firewall link color in NodeBalancer summary is the correct color. ✅ 

I also tested add/remove functionality from the various flows and didn't find any issues.

File: packages/manager/src/features/Events/eventMessageGenerator.ts [hunk 19]
```diff
@@ -29,6 +29,16 @@ export const safeSecondaryEntityLabel = (
   return label ? `${text} ${label}` : fallback;
 };
 
+interface SecondaryEntityType {
```
Comment (dwiley-akamai) [inline packages/manager/src/features/Events/eventMessageGenerator.ts:32]: Since this seems to be used to point out what entity types can be in that `e.secondary_entity` field, I wonder if this would be better as an actual `type`, i.e. `type SecondaryEntityType = 'linode' | 'nodebalancer'`

File: packages/manager/src/features/NodeBalancers/NodeBalancerDetail/NodeBalancerSummary/SummaryPanel.tsx [hunk 20]
```diff
@@ -90,6 +95,18 @@ export const SummaryPanel = () => {
...
+          <StyledTitle data-qa-title variant="h3">
+            Firewall
+          </StyledTitle>
+          <Typography data-qa-firewall variant="body1">
+            <Link className="secondaryLink" to={`/firewalls/${linkID}`}>
+              {linkText}
+            </Link>
+          </Typography>
+        </StyledSummarySection>
+      ) : null}
```
Comment (dwiley-akamai) [inline packages/manager/src/features/NodeBalancers/NodeBalancerDetail/NodeBalancerSummary/SummaryPanel.tsx:109]: You could use another `&&` here to get rid of the ternary if you wish

File: packages/manager/src/features/NodeBalancers/NodeBalancerDetail/NodeBalancerSettings.tsx [hunk 21]
```diff
@@ -5,22 +5,29 @@ import { useParams } from 'react-router-dom';
...
 
 export const NodeBalancerSettings = () => {
+  const flags = useFlags();
   const theme = useTheme();
   const { nodeBalancerId } = useParams<{ nodeBalancerId: string }>();
   const id = Number(nodeBalancerId);
-
   const { data: nodebalancer } = useNodeBalancerQuery(id);
+  const { data: attachedFirewallData } = useNodeBalancersFirewallsQuery(id);
+  const displayFirewallInfoText = attachedFirewallData?.data?.length
```
Comment (dwiley-akamai) [inline packages/manager/src/features/NodeBalancers/NodeBalancerDetail/NodeBalancerSettings.tsx:28]: What if we did

`const displayFirewallInfoText = Boolean(attachedFirewallData?.data?.length);`

instead?

File: packages/manager/src/features/NodeBalancers/NodeBalancerDetail/NodeBalancerFirewalls.tsx [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+        firewall={attachedFirewall}
+        key={`firewall-${attachedFirewall.id}`}
+        nodeBalancerID={nodeBalancerId}
+        onClickUnassign={handleClickUnassign}
+      />
+    ));
+  };
+
+  const learnMoreLink = (
+    <Link to="https://www.linode.com/docs/products/networking/cloud-firewall/guides/create-a-cloud-firewall/">
```
Comment (dwiley-akamai) [inline packages/manager/src/features/NodeBalancers/NodeBalancerDetail/NodeBalancerFirewalls.tsx:80]: This link gets used in a couple of places, how about making it a constant somewhere?

**Review by dwiley-akamai — APPROVED** [review]
Functionality and code LGTM now, left a few fairly minor comments

File: packages/manager/src/features/Events/eventMessageGenerator.ts [hunk 23]
```diff
@@ -29,6 +29,16 @@ export const safeSecondaryEntityLabel = (
...
 
+type SecondaryEntityType = {
+  linode: string;
+  nodebalancer: string;
+};
+
+const secondaryEntityTypeObj: SecondaryEntityType = {
+  linode: 'Linode',
+  nodebalancer: 'NodeBalancer',
+};
```
Comment (bnussman-akamai) [inline packages/manager/src/features/Events/eventMessageGenerator.ts:40]:
```suggestion
const secondaryFirewallEntityNameMap: Record<
  FirewallDeviceEntityType,
  string
> = {
  linode: 'Linode',
  nodebalancer: 'NodeBalancer',
}
```

File: packages/manager/src/features/Firewalls/FirewallLanding/CreateFirewallDrawer.tsx [hunk 24]
```diff
@@ -169,24 +243,70 @@ export const CreateFirewallDrawer = React.memo(
...
+                margin: `${theme.spacing(2)} ${theme.spacing(0)}`,
+              })}
+              variant="h3"
+            >
+              {FirewallLabelText}
+            </Typography>
+            <Typography>
+              {FirewallHelperText}
+              {deviceSelectGuidance ? ` ${deviceSelectGuidance}` : null}
+            </Typography>
```
Comment (coliu-akamai) [inline packages/manager/src/features/Firewalls/FirewallLanding/CreateFirewallDrawer.tsx:261]: another accessibility thought - If it's within accessibility guidelines (will need to read up on them more since I'm not too familiar with them yet 😅), could we add aria-labels to the header and helper text of the create firewall drawer so that they're read when using voiceover?

image

currently only the link 'Learn more' is read, but not 'assign X to the Firewall' or the helper text, which could be potentially confusing to screenreaders

**Review by coliu-akamai — APPROVED** [review]
🎉  awesome job! just a couple small comments

✅ went over functionality
✅ rest of code review

**Review by bnussman-akamai — COMMENTED** [review]
Only found two potential issues

Dark mode color issue on the NodeBalancer's Settings page
Screenshot 2023-12-07 at 2 11 51 PM

Looks like we may not be invalidating entity firewall queries upon deletion. Do we need to do this? If so, let me know if you need help with it!
https://github.com/linode/manager/assets/115251059/5cecfbf7-c103-4fde-aa8e-a297e025f05b

File: packages/manager/src/utilities/mapIdsToDevices.test.ts [hunk 25]
```diff
@@ -1,22 +1,22 @@
 import { linodeFactory } from 'src/factories';
 
-import { mapIdsToLinodes } from './mapIdsToLinodes';
+import { mapIdsToDevices } from './mapIdsToDevices';
 
-describe('mapIdsToLinodes', () => {
+describe('mapIdsToDevices', () => {
```
[the remark this answers is not in the document]
Comment (tyler-akamai) [reply packages/manager/src/utilities/mapIdsToDevices.test.ts:9]: Just added NodeBalancer coverage

**Comment by carrillo-erik** [comment]
> Dark mode color issue on the NodeBalancer's Settings page Screenshot 2023-12-07 at 2 11 51 PM
> 

Fixed the dark mode color issue.

Screenshot 2023-12-07 at 12 01 02 PM

File: packages/manager/src/features/Firewalls/FirewallLanding/CreateFirewallDrawer.tsx (hunk 24 above)
Comment (carrillo-erik) [reply packages/manager/src/features/Firewalls/FirewallLanding/CreateFirewallDrawer.tsx:261]: @coliu-akamai 
Screenshot 2023-12-07 at 12 23 38 PM

You can read more about it here:
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-label

Which browser and screenreader are you using? When I turn on voiceover on my machine and using Chrome, the header and helper text are read.

**Comment by tyler-akamai** [comment]
sorry accidentally closed the pr, just reopened it

**Comment by tyler-akamai** [comment]
> Looks like we may not be invalidating entity firewall queries upon deletion. Do we need to do this? If so, let me know if you need help with it!

Invalidated entity firewall queries upon deletion.

File: packages/manager/src/features/Firewalls/FirewallLanding/FirewallDialog.tsx [hunk 26]
```diff
@@ -1,42 +1,51 @@
...
   const {
     mode,
     onClose,
     open,
-    selectedFirewallID,
+    selectedFirewallId,
     selectedFirewallLabel: label,
   } = props;
 
+  const { data: devices } = useAllFirewallDevicesQuery(
```
Comment (tyler-akamai) [inline packages/manager/src/features/Firewalls/FirewallLanding/FirewallDialog.tsx:35]: I didnt add any error handling because:
1. Its not one of the three main actions
2. And if it does fail to invalidate the queries, the queries will be updated next refresh

File: packages/manager/src/features/Firewalls/FirewallLanding/CreateFirewallDrawer.tsx (hunk 24 above)
Comment (coliu-akamai) [reply packages/manager/src/features/Firewalls/FirewallLanding/CreateFirewallDrawer.tsx:261]: awesome ty!
